vignettes/data_structures.Rmd

```diff
@@ -0,0 +1,122 @@
+was computed from `line1` and `line2` respectively.
+
+So far, styler can set the spaces around the operators correctly. In our example, 
+that involves adding spaces around `+`, so in the `spaces` column, element nine
+and ten must be set to one. This means that a space is added after `1` and after `+`. 
+To get the spacing right and cover the various cases, a set of functions has to 
+be applied to the parse table subsequently (and in the right order), 
+which is essentially done via `Reduce()`. 
+After all modifications on the table are completed, `serialize_parse_data()`
+collapses the `text` column and adds the number of spaces and 
+line breaks specified in `spaces` and `newlines` in between the elements of
+`text`. If we serialize our table and don't perform any modification, we 
+obviously just get back what we started with.
+```{r}
+serialize_parse_data(parse_table)
+```
+
+## Refining the flat approach - nesting the parse table
+
+Although the flat approach is good place to start, e.g. for fixing spaces
+between operators, it has its limitations. In particular, it treats each token 
+the same way in the sense that it does not account for the context of the token, 
+i.e. in which sub-expression it appears.
+To set the indention correctly, we need a hierarchical view on the parse data, 
+since all tokens in a sub-expression have the same indention level. Hence, 
+a natural approach would be to create a nested parse table instead of a flat
+parse table and than take a recursion over all elements in the table, so for 
+each sub(-sub etc.)-expression, a separate parse table would be created and the 
+modifications would be applied to this table before putting everything back 
+together. A function to create a nested parse table already exists in `styler`.
+Let's have a look at the top level:
+
+```{r}
+(l1 <- compute_parse_data_nested(code)[-1])
+
+```
+
+The tibble contains the column `leaves`, which itself contains a tibble. 
+If we "enter"  the first leaf, we can see that the expression was split up 
+further.
+
+```{r}
+l1$leaves[[1]] %>%
+  select(text, terminal, leaves, token)
+```
+
+And further...
+```{r}
+l1$leaves[[1]]$leaves[[3]]$leaves[[5]]$leaves[[2]]
+```
+
+... and so on. Every leaf that is not a terminal contains another tibble where 
+the sub-expression is split up further - until we are left with tibbles that 
+only contain terminals.
+
+
+Recall the above example. `1+1` 
+is a sub-expression of the expression `code`, and we would like to set the 
+indention for `1+1`, even before we do anything with `1+1`.
+
+The code for a function that kind of sketches the idea and illustrates such a 
+recursion is given below: 
+
+```{r}
+serialize <- function(x) {
+  out <- Map(
+    function(terminal, token, leaf) {
+      if (terminal)
+        token
+      else
+        serialize(leaf)
+    },
+    x$terminal, x$token,
+    if (is.null(x$leaves)) rep(list(NULL), length(x$terminal)) else x$leaves
```

Comment:
No problem, but the code can be made shorter if you just use `x$child` instead of this `if` expression.